File: ios/Packages/AppFeatures/Sources/SearchHistoryFeature/SearchHistoryRecorder.swift
import CoreKit
import Foundation
import NetworkingKit

/// Concrete `SearchHistoryRecording` (CoreKit) — the thing Dua (and any future
/// searchable feature) is injected with. Fire-and-forget: failures are silent
/// (docs/features/search-history.md), never surfaced to the search UI.
public final class SearchHistoryRecorder: SearchHistoryRecording {
    private let client: AuthenticatedAPIClientProtocol

    public init(client: AuthenticatedAPIClientProtocol) {
        self.client = client
    }

    public func record(source: String, queryText: String, locale: String) {
        Task { [client] in
            let _: SearchHistoryEntry? = try? await client.post(
                "v1/search-history",
                body: RecordSearchRequest(source: source, query_text: queryText, locale: locale)
            )
        }
    }
}

File: android/feature/searchhistory/src/main/kotlin/com/fatwabot/feature/searchhistory/SearchHistoryRecorder.kt
package com.fatwabot.feature.searchhistory

import com.fatwabot.core.common.SearchHistoryRecording
import com.fatwabot.core.network.AuthenticatedApiClientProtocol
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json

/** Concrete SearchHistoryRecording — the thing Dua (and any future searchable
 * feature) is injected with. Fire-and-forget: failures are silent
 * (docs/features/search-history.md), never surfaced to the search UI. */
class SearchHistoryRecorder(
    private val client: AuthenticatedApiClientProtocol,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.IO),
) : SearchHistoryRecording {
    private val json = Json { ignoreUnknownKeys = true }

    override fun record(source: String, queryText: String, locale: String) {
        scope.launch {
            runCatching {
                val body = json.encodeToString(
                    RecordSearchRequest.serializer(),
                    RecordSearchRequest(source = source, queryText = queryText, locale = locale),
                )
                client.postRaw("v1/search-history", body)
            }
        }
    }
}
